File: renderers/android/src/main/kotlin/dev/varn/gui/VarnSurfaceViews.kt
package dev.varn.gui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Shader
import org.json.JSONArray

/** A box painted with a run of colours, which is what a screen with a header of its own is built on. */
class VarnGradientView(context: Context) : VarnBoxView(context) {
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var colours = intArrayOf()
    private var stops: FloatArray? = null
    private var direction = "down"

    init {
        setWillNotDraw(false)
    }

    fun setColors(values: JSONArray?) {
        val read = mutableListOf<Int>()

        for (index in 0 until (values?.length() ?: 0)) {
            read += VarnStyle.color(values?.optString(index)) ?: continue
        }

        colours = read.toIntArray()
        shade()
    }

    fun setLocations(values: JSONArray?) {
        if (values == null) {
            stops = null
            shade()
            return
        }

        stops = FloatArray(values.length()) { index -> values.optDouble(index).toFloat() }
        shade()
    }

    fun setDirection(value: String?) {
        direction = value ?: "down"
        shade()
    }

    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
        super.onSizeChanged(width, height, oldWidth, oldHeight)
        shade()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        if (colours.size < 2 || width == 0 || height == 0) {
            return
        }

        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)
    }

    private fun shade() {
        if (colours.size < 2 || width == 0 || height == 0) {
            return
        }

        val wide = width.toFloat()
        val tall = height.toFloat()

        val (from, to) = when (direction) {
            "up" -> (0f to tall) to (0f to 0f)
            "right" -> (0f to 0f) to (wide to 0f)
            "left" -> (wide to 0f) to (0f to 0f)
            "diagonal" -> (0f to 0f) to (wide to tall)
            else -> (0f to 0f) to (0f to tall)
        }

        paint.shader = LinearGradient(
            from.first, from.second, to.first, to.second,
            colours, stops, Shader.TileMode.CLAMP,
        )

        invalidate()
    }
}

/**
 * A box that stands over what is behind it, which is what a bar over a scrolling screen is made of.
 *
 * Android has no blur of a view's own backdrop, so what it draws is the tint at the weight the tree
 * asked for. iOS and the browser blur what is behind the same box, which is why the tint is what
 * carries the colour rather than the blur carrying it.
 */
class VarnBlurView(context: Context) : VarnBoxView(context) {
    private var intensity = 0.85f
    private var tint = Color.TRANSPARENT

    fun setIntensity(value: Float) {
        intensity = value.coerceIn(0f, 1f)
        wash()
    }

    fun setTint(value: String?) {
        tint = VarnStyle.color(value) ?: Color.TRANSPARENT
        wash()
    }

    private fun wash() {
        val alpha = (Color.alpha(tint) * intensity).toInt()

        setBackgroundColor(Color.argb(alpha, Color.red(tint), Color.green(tint), Color.blue(tint)))
    }
}

File: renderers/ios/VarnSurfaceViews.swift
import UIKit

/// A box painted with a run of colours, which is what a screen with a header of its own is built on.
final class VarnGradientView: UIView {
    override class var layerClass: AnyClass { CAGradientLayer.self }

    private var painted: [CGColor] = []
    private var stops: [NSNumber]?

    private var gradient: CAGradientLayer {
        layer as! CAGradientLayer
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        setDirection("down")
    }

    required init?(coder: NSCoder) { fatalError("not used") }

    func setColors(_ values: [Any]) {
        painted = values.compactMap { VarnStyle.color($0) }.map(\.cgColor)
        gradient.colors = painted
    }

    func setLocations(_ values: [Any]?) {
        stops = values?.compactMap { VarnValue.number($0).map { NSNumber(value: $0) } }
        gradient.locations = stops
    }

    /// Points the run of colour, which is one of a fixed set so the angle is the same everywhere.
    func setDirection(_ value: String?) {
        switch value {
        case "up":
            gradient.startPoint = CGPoint(x: 0.5, y: 1)
            gradient.endPoint = CGPoint(x: 0.5, y: 0)
        case "right":
            gradient.startPoint = CGPoint(x: 0, y: 0.5)
            gradient.endPoint = CGPoint(x: 1, y: 0.5)
        case "left":
            gradient.startPoint = CGPoint(x: 1, y: 0.5)
            gradient.endPoint = CGPoint(x: 0, y: 0.5)
        case "diagonal":
            gradient.startPoint = CGPoint(x: 0, y: 0)
            gradient.endPoint = CGPoint(x: 1, y: 1)
        default:
            gradient.startPoint = CGPoint(x: 0.5, y: 0)
            gradient.endPoint = CGPoint(x: 0.5, y: 1)
        }
    }

    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        VarnHit.through(super.hitTest(point, with: event), self)
    }
}

/// A box that blurs the screen behind it, which is what a bar over a scrolling page is made of.
///
/// The blur is the system's own material, so it follows the appearance the way every other one does,
/// and the tint is drawn over it for a bar that carries a colour of its own.
final class VarnBlurView: UIView {
    private let effect = UIVisualEffectView(effect: UIBlurEffect(style: .systemMaterial))
    private let content = VarnContentView()
    private let wash = UIView()

    override init(frame: CGRect) {
        super.init(frame: frame)

        clipsToBounds = true
        addSubview(effect)
        addSubview(wash)
        addSubview(content)
    }

    required init?(coder: NSCoder) { fatalError("not used") }

    /// Answers the view children are added to, which is above the material rather than inside it.
    var contentView: UIView { content }

    func setIntensity(_ value: CGFloat) {
        effect.alpha = min(1, max(0, value))
    }

    func setTint(_ value: String?) {
        wash.backgroundColor = VarnStyle.color(value)
    }

    override func layoutSubviews() {
        super.layoutSubviews()

        effect.frame = bounds
        wash.frame = bounds
        content.frame = bounds
    }

    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        VarnHit.through(super.hitTest(point, with: event), self)
    }
}

/// A spinner drawn at the size the engine gave it rather than at one of the two the platform has.
///
/// `UIActivityIndicatorView` draws at its own intrinsic size whatever bounds it is put in, so a box of
/// thirty-two points held a twenty point spinner and the size above it drew no larger. It is scaled
/// inside a view of its own, since the node's own transform is written from its style on every commit
/// and anything set on the node would be wiped by the next one.
final class VarnActivityView: UIView {
    private let indicator = UIActivityIndicatorView(style: .medium)

    override init(frame: CGRect) {
        super.init(frame: frame)

        addSubview(indicator)
        indicator.startAnimating()
    }

    required init?(coder: NSCoder) { fatalError("not used") }

    var animating: Bool {
        get { indicator.isAnimating }

        set {
            if newValue {
                indicator.startAnimating()
                return
            }

            indicator.stopAnimating()
        }
    }

    var color: UIColor? {
        get { indicator.color }
        set { indicator.color = newValue }
    }

    override var intrinsicContentSize: CGSize {
        indicator.intrinsicContentSize
    }

    override func layoutSubviews() {
        super.layoutSubviews()

        let side = min(bounds.width, bounds.height)
        let drawn = indicator.intrinsicContentSize.width

        indicator.transform = .identity
        indicator.center = CGPoint(x: bounds.midX, y: bounds.midY)

        guard drawn > 0, side > 0 else {
            return
        }

        indicator.transform = CGAffineTransform(scaleX: side / drawn, y: side / drawn)
    }
}
